snapshot/snapshot.go

```diff
@@ -335,24 +344,27 @@ func (o *snapshotter) Prepare(ctx context.Context, key, parent string, opts ...s
 		// don't fallback here, since there was an error getting mounts
 		return nil, err
 	}
-
-	log.G(ctx).WithField("layerDigest", base.Labels[ctdsnapshotters.TargetLayerDigestLabel]).Info("preparing snapshot as local snapshot")
-	err = o.prepareLocalSnapshot(lCtx, key, base.Labels, mounts)
-	if err == nil {
-		err := o.commit(ctx, false, target, key, append(opts, snapshots.WithLabels(base.Labels))...)
-		if err == nil || errdefs.IsAlreadyExists(err) {
-			// count also AlreadyExists as "success"
-			// there's no need to provide any details on []mount.Mount because mounting is already taken care of
-			// by snapshotter
-			log.G(lCtx).Info("local snapshot successfully prepared")
-			return nil, fmt.Errorf("target snapshot %q: %w", target, errdefs.ErrAlreadyExists)
+	// If the underlying FileSystem deems that the image is unable to be lazy loaded, then we
+	// should completely fallback to the underlying runtime (containerd) to handle pulling and
+	// unpacking all the layers in the image.
+	if !skipLazyLoadingImage {
+		log.G(ctx).WithField("layerDigest", base.Labels[ctdsnapshotters.TargetLayerDigestLabel]).Info("preparing snapshot as local snapshot")
+		err = o.prepareLocalSnapshot(lCtx, key, base.Labels, mounts)
+		if err == nil {
+			err := o.commit(ctx, false, target, key, append(opts, snapshots.WithLabels(base.Labels))...)
+			if err == nil || errdefs.IsAlreadyExists(err) {
+				// count also AlreadyExists as "success"
+				// there's no need to provide any details on []mount.Mount because mounting is already taken care of
+				// by snapshotter
+				log.G(lCtx).Info("local snapshot successfully prepared")
+				return nil, fmt.Errorf("target snapshot %q: %w", target, errdefs.ErrAlreadyExists)
+			}
+			log.G(lCtx).WithError(err).Warn("failed to internally commit local snapshot")
+			// Don't fallback here (= prohibit to use this key again) because the FileSystem
+			// possible has done some work on this "upper" directory.
+			return nil, err
```

Comment:
Might be worth refactoring this section to reduce cyclomatic complexity. For example, 

```
if skipLazyLoading {
    return mounts, nil
}

// Continue to handle edge cases
...
```